Top Posts Widget: Enable transient caching of results to reduce DB calls.

georgestephanis opened the pull request:
In my testing, using the widget on a page on a client site has resulted in over 2,000 additional db queries, based on the loop to `get_attached_media()` for every potential top post firing off numerous subsequent db queries --

```
830 are querying wp_posts to do WP_Post::get_instance()
878 are update_meta_cache()
517 are WP_Term_Query->get_terms()
133 are WP_Query->get_posts()
```

## Proposed changes:
* Add transient caching for the top posts.

## Testing instructions:
* Load a page with the widget and Query Monitor -- with a lot of past traffic.
* See the queries go down?

## Does this pull request change what data or activity we track or use?

Nope

georgestephanis committed:
Use WP Constants instead of ✨ magic numbers ✨

georgestephanis committed:
Add transient caching for the top posts.

In my testing, using the widget on a page on a client site has resulted in over 2,000 additional db queries, based on the loop to `get_attached_media()` for every potential top post firing off numerous subsequent db queries --

830 are querying wp_posts to do WP_Post::get_instance()
878 are update_meta_cache()
517 are WP_Term_Query->get_terms()
133 are WP_Query->get_posts()

jeherve commented on projects/plugins/jetpack/_inc/lib/class-jetpack-top-posts-helper.php:
```diff
@@ -137,6 +144,10 @@ function ( $item ) use ( $acceptable_types ) {
 		}
 
+		if ( $cached ) {
+			set_transient( $transient_key, $top_posts, DAY_IN_SECONDS );
```
A day may be a bit too long, folks will expect top posts to be updated after they publish a post, and if that post immediately gets some traffic from social media for example. 

Maybe we could use the same expiration as when fetching the top posts from WordPress.com?
https://github.com/Automattic/jetpack/blob/e9a49af5ddd83d2a0e282178f2b11e9fdfcbb8f5/projects/packages/stats/src/class-wpcom-stats.php#L455-L468

It's currently set to 5 minutes by default but can be overwritten with the filter.

georgestephanis replied:
I'm fine with adjusting it to whenever -- I'm just hoping to have it ease up on the db load on every page view.

georgestephanis replied:
```suggestion
			$minutes = class_exists( 'WPCOM_Stats' ) ? WPCOM_Stats::STATS_CACHE_EXPIRATION_IN_MINUTES : 5;
			set_transient( $transient_key, $top_posts, $minutes * MINUTE_IN_SECONDS  );
```

If we wanted to safely rely on the `WPCOM_Stats` class's constant with a fallback.  I'm not sure if there's a situation where it wouldn't be available.

georgestephanis replied:
My only real concern there is what frequency of churn we'd want to go into the options table, as it would be db writes on the front-end if there isn't a storage for transients in memcached or redis or the like.

jeherve replied:
> If we wanted to safely rely on the WPCOM_Stats class's constant with a fallback. I'm not sure if there's a situation where it wouldn't be available.

What do you think about creating a new public method in `WPCOM_Stats`, that would return `self::STATS_CACHE_EXPIRATION_IN_MINUTES * MINUTE_IN_SECONDS` wrapped in the existing `jetpack_fetch_stats_cache_expiration` filter? You'd then be able to call that from `Jetpack_Top_Posts_Helper` (no need for the `class_exists` check, the class is autoloaded) and you'd be able to customize the expiration from a functionality plugin on your client's side if 5 minutes is still too much.

georgestephanis replied:
Sure, sounds reasonable.

jeherve reviewed (commented):
Hey George! Thanks for the patch. That seems like a good idea.

I have a suggestion below.

Do you think you could also create a changelog entry for that change, with `cd projects/plugins/jetpack && composer install &$ vendor/bin/changelogger add`?

georgestephanis committed:
Create fix-top-posts-transient-caching

georgestephanis commented:
Once I'd set up the build tools locally, the desktop GitHub app suddenly was kicking errors when I tried to push the commit with the changelog entry --

<img width="697" alt="Screenshot 2025-05-29 at 12 28 04 PM" src="https://github.com/user-attachments/assets/87d900cf-fb41-4c40-baa3-89a0cc5d0e1d" />

I wound up manually creating the changelog entry through the github.com code editor, but noting here in case it's indicative of a broader issue that could make it more difficult for folks to contribute.

kangzj commented:
Nice improvement @georgestephanis. Thanks!

> Interestingly, while the block uses get_attached_media, the widget uses Jetpack_PostImages::get_image(). In a follow-up PR it may be interesting to update the widget to rely on the same Jetpack_Top_Posts_Helper class. @Automattic/red This may be an interesting project to improve the performance and consistency of the feature, maybe?

It does feel like something that's better to be consistent. I'll add a task but I guess it's not high priority.

cc @Aurorum

Aurorum commented:
For context, if I remember correctly, I think that `Jetpack_Top_Posts_Helper` isn’t able to cover popularity based on Likes whereas the widget does, which is why I initially just used it for the blocks (the class prevents duplication of the code in the block’s preview and output). In principle though, it seems sensible, so will revisit to see if anything comes to mind or has changed!

georgestephanis committed:
Drop Serialize for json encoding, and don't treat $types as an array

(it's a string -- whoops!)

Also, split out the assignment to appease static code analysis deities.

georgestephanis commented:
Confirming that implementing the nuts and bolts of this on the client's site has resulted in a HUGE performance improvement, as it was running on pretty much every page load previously.

georgestephanis commented:
Noting that the failing check was based on getting the token -- doesn't seem to be related to the content of the PR

jeherve commented:
@georgestephanis Did you have the chance to look into https://github.com/Automattic/jetpack/pull/43688#discussion_r2115678560 ?
